### PR TITLE
build(docs-infra): fail on unused doc-regions

### LINKED ISSUE #19761: dgeni should be able to find unused docregions and report it
There are some docregions in the docs that are not used anywhere.

### PR DESCRIPTION
**Pull request by petebacondarwin**
(I guess it is likely that this will not backport cleanly to patch)

Fixes #19761

### CODE REVIEW

File: aio/content/examples/testing/src/app/demo/demo.testbed.spec.ts [hunk 1]
```diff
@@ -628,7 +617,7 @@ describe('demo (with TestBed):', () => {
 
       child.childValue = 'bar';
 
-      return new Promise(resolve => {
+      return new Promise<void>(resolve => {
```
Comment (gkalpak) [inline aio/content/examples/testing/src/app/demo/demo.testbed.spec.ts:620]: This seems unrelated to this commit 🤔

File: packages/compiler/src/url_resolver.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@ export function createOfflineCompileUrlResolver(): UrlResolver {
  *
  * ## Example
  *
- * {@example compiler/ts/url_resolver/url_resolver.ts region='url_resolver'}
+ * <code-example path="compiler/ts/url_resolver/url_resolver.ts"></code-example>
```
Comment (gkalpak) [inline packages/compiler/src/url_resolver.ts:29]: Why remove the region here?

File: packages/examples/compiler/ts/url_resolver/url_resolver.ts [hunk 3]
```diff
@@ -15,7 +15,6 @@ import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
 class MyApp {
 }
 
-// #docregion url_resolver
```
Comment (gkalpak) [inline packages/examples/compiler/ts/url_resolver/url_resolver.ts:18]: Why remove this docregion (since it was used in [packages/compiler/src/url_resolver.ts][1] before this PR)?

[1]: https://github.com/angular/angular/blob/0d51a81a44d58cbdcecdf73e60dc15c9e48ca364/packages/compiler/src/url_resolver.ts#L29

**Review by gkalpak — CHANGES_REQUESTED** [review]
:100: :100: :100:

File: aio/tools/transforms/authors-package/utils.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+module.exports = {
+  codeExampleMatcher: /<code-(?:pane|example)(?:[^>]|[ \r\n])*path="([^"]+)"/g,
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/authors-package/utils.js:10]: The original regex did not match examples that covered multiple lines, e.g.

```
<code-example
  path="styleguide/src/05-03/app/heroes/shared/hero-button/hero-button.component.avoid.ts"
  region="example"
  header="app/heroes/hero-button/hero-button.component.ts">
</code-example>
```

I thought it was because we were not picking up the newlines, but now I see it was because the original regex required a space character after the element name. 🤦

File: aio/tools/transforms/angular-base-package/index.js [hunk 5]
```diff
@@ -53,10 +53,11 @@ module.exports = new Package('angular-base', [
     inlineTagProcessor.inlineTagDefinitions.push(require('./inline-tag-defs/custom-search-defs/'));
   })
 
-  .config(function(checkAnchorLinksProcessor) {
-    // This is disabled here to prevent false negatives for the `docs-watch` task.
+  .config(function(checkAnchorLinksProcessor, checkForUnusedExampleRegions) {
+    // These are disabled here to prevent false negatives for the `docs-watch` task.
     // It is re-enabled in the main `angular.io-package`
     checkAnchorLinksProcessor.$enabled = false;
+    checkForUnusedExampleRegions.$enabled = false;
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/index.js:60]: The trouble is that there are multiple authoring packages and we need to remember to disable it in each of them. So it felt more dry to do it this way round.
But perhaps we could refactor it so that all authoring packages extended a standard one that did these shared disablings.
I'll add an extra commit.

File: aio/tools/transforms/examples-package/processors/check-for-unused-example-regions.spec.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            '': {id: 'some/file#'},
+            'used': {id: 'some/file#used', usedInDoc: {}},
+            'not-used': {id: 'some/file#not-used'},
+          }
+        }
+      };
+      expect(() => processor.$process())
+          .toThrowError('There is 1 unused example region:\n - some/file#not-used');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/examples-package/processors/check-for-unused-example-regions.spec.js:35]: You assume correct!

File: aio/content/examples/testing/src/app/demo/demo.testbed.spec.ts (hunk 1 above)
Comment (petebacondarwin) [reply aio/content/examples/testing/src/app/demo/demo.testbed.spec.ts:620]: Well yes, but my IDE was saying that the file was invalid without this change.
So I thought I would just sneak it in :-D
I'll move it to its own commit.

File: packages/examples/compiler/ts/url_resolver/url_resolver.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/examples/compiler/ts/url_resolver/url_resolver.ts:18]: My mistake!

Comment (petebacondarwin) [reply packages/examples/compiler/ts/url_resolver/url_resolver.ts:18]: Actually, I did mean to do this.  The thing is that the `UrlResolver` in `packages/compiler/src/url_resolver.ts` is not public API. So the example there is not used in the docs. The result of this is that the doc-region here causes a failure, since it is not "used".

So I removed it. But then I thought, it doesn't make sense to reference the region in the non-public API doc, so I changed that to use the unnamed region, which would still work if that did become part of the public API.

File: packages/compiler/src/url_resolver.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler/src/url_resolver.ts:29]: See https://github.com/angular/angular/pull/40479#discussion_r560854939

File: aio/tools/transforms/authors-package/marketing-package.js [hunk 7]
```diff
@@ -7,10 +7,11 @@
  */
 const Package = require('dgeni').Package;
 const contentPackage = require('../angular-content-package');
+const baseAuthoringPackage = require('./base-authoring-package');
 const { CONTENTS_PATH } = require('../config');
```
Comment (gkalpak) [inline aio/tools/transforms/authors-package/marketing-package.js:11]: 😇 

```suggestion
const {CONTENTS_PATH} = require('../config');
const baseAuthoringPackage = require('./base-authoring-package');
```

File: aio/tools/transforms/authors-package/base-authoring-package.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    .config(function(linkInlineTagDef) {
+      // Do not fail the processing if there is an invalid link
+      linkInlineTagDef.failOnBadLink = false;
+    })
+
+    .config(function(renderExamples) {
+      // Do not fail the processing if there is a broken example
+      renderExamples.ignoreBrokenExamples = true;
+    });
```
Comment (gkalpak) [inline aio/tools/transforms/authors-package/base-authoring-package.js:34]: OOC, why the separate `config()` blocks for `linkInlineTagDef` and `renderExamples`, but not for `checkAnchorLinksProcessor` and `checkForUnusedExampleRegions`?

File: aio/tools/transforms/angular-base-package/index.js [hunk 9]
```diff
@@ -5,17 +5,21 @@
...
 const gitPackage = require('dgeni-packages/git');
 const jsdocPackage = require('dgeni-packages/jsdoc');
 const nunjucksPackage = require('dgeni-packages/nunjucks');
+const postProcessPackage = require('dgeni-packages/post-process-html');
+
+const { SRC_PATH } = require('../config');
 const linksPackage = require('../links-package');
 const examplesPackage = require('../examples-package');
 const targetPackage = require('../target-package');
 const remarkPackage = require('../remark-package');
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/index.js:22]: Since you're ordering imports alphabetically :grin:

```suggestion
const remarkPackage = require('../remark-package');
const targetPackage = require('../target-package');
```

File: packages/examples/compiler/ts/url_resolver/url_resolver.ts (hunk 3 above)
Comment (gkalpak) [reply packages/examples/compiler/ts/url_resolver/url_resolver.ts:18]: I see. That's not ideal, but if it's only one occurrence not worth trying to fix it "properly"(:tm:) :grin:

File: aio/tools/transforms/authors-package/base-authoring-package.js (hunk 8 above)
Comment (petebacondarwin) [reply aio/tools/transforms/authors-package/base-authoring-package.js:34]: Generally I like to keep the config blocks focussed on a single task. In this case the first block is disabling two processors. In the other blocks we are configuring different aspects of other injectables. They could indeed all go in one, but I have found that this leads to people bundling lots of stuff into a single block, so I like to start out as I wish to carry on. Equally the first config block could have been split into two but then I would have had to duplicate the comment :-D

File: packages/examples/compiler/ts/url_resolver/url_resolver.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/examples/compiler/ts/url_resolver/url_resolver.ts:18]: It might be possible to avoid this problem by moving the `checkForUnusedExampleRegions` process earlier in the processing before we remove non-public APIs. But I have already spent more time than I had planned on this, so I am loathe to start further investigations.

**Review by gkalpak — APPROVED** [review]
Great work :tada: I love the clean-up :heart:
(All these unused docregions have been bugging me for a long time :grin:)

**Review by gkalpak — APPROVED** [review]
Reviewed-for: global-docs-approvers

**Comment by petebacondarwin** [comment]
Caretaker: This PR is a docs/docs-infra only change so a presubmit should not be required.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

Note: from `public-api` perspective there are just 3 files in this folder: `aio/content/examples/styleguide`.

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Review by atscott — APPROVED** [review]
reviewed-for: public-api
